### PR TITLE
Revise @ in patterns section

### LINKED ISSUE #25008: @-pattern documentation in the book is a little misleading
http://doc.rust-lang.org/book/patterns.html#bindings . Current text here for future triage purposes:

> # Bindings

> If you’re matching multiple things, via a | or a ..., you can bind the value to a name with @:

> ```rust
> let x = 1;
>
> match x {
>     e @ 1 ... 5 => println!("got a range element {}", e),
>     _ => println!("anything"),
> }
> ```

> This prints `got a range element 1`.

The `@` binding is more general than just `|` or `...` patterns: it can be attached to literally any pattern, e.g.

```rust
let mut x: Option<SomeStruct> = ...;
match x {
    Some(SomeStruct { a_field: ref a @ Some(_), .. }) => {}
    Some(ref mut b @ SomeStruct { something_else: 5, .. }) => {},
    Some(SomeStruct { third_one: c @ &Ok(_) }) => {},
    _ => {}
}
```

Also, for `|`, the `@` has to be attached to each alternative:

```rust
let x = 1;
match x {
    y @ 1...5 | y @ 10...15 => println!("{}", y),
    _ => {}
}
```

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #25008

### CODE REVIEW

File: src/doc/trpl/patterns.md [hunk 1]
```diff
@@ -82,7 +81,36 @@ match x {
...
+This prints `Some("Steve")`: We’ve bound the inner `name` to `a`.
+
+If you use `@` with `|`, you need to make sure the name is bound in each part
+of the pattern:
+
+```rust
+let x = 5;
+
+match x {
+    e @ 1 ... 5 | e @ 2 ... 6 => println!("got a range element {}", e),
```
Comment (huonw) [inline src/doc/trpl/patterns.md]: Might be more illustrative to have the ranges be disjoint? (e.g. the second one could start at 8)

**Comment by huonw** [comment]
r=me with the range comment.

**Comment by steveklabnik** [comment]
(forgot to actually commit the change)

File: src/doc/trpl/patterns.md [hunk 2]
```diff
@@ -70,8 +70,7 @@ This prints `something else`
 
 # Bindings
 
-If you’re matching multiple things, via a `|` or a `...`, you can bind
-the value to a name with `@`:
+You can bind values to names with `@`:
```
Comment (jakub-) [inline src/doc/trpl/patterns.md]: Hmm, could we be more elaborate here, maybe? The point of `@` is that you can both bind the whole value and destructure it in the same pattern. If one only wanted to bind the value to a name, they could use an identifier.

See what I mean? I can't quite put it into words myself. :-)

Comment (jakub-) [inline src/doc/trpl/patterns.md]: @steveklabnik I know this is merged but I wonder if you the above is worth addressing separately.

**Pull request merged**